app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -44,7 +44,7 @@ echo "#"
 if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     register_katello_host(){
         UUID=$(subscription-manager identity | head -1 | awk '{print $3}')
-        curl --request POST "<%= foreman_server_url %>/register?uuid=$UUID" \
+        curl --request POST "<%= @registration_url %>?uuid=$UUID" \
```

Comment:
If you pass in a URL that includes get parameters, things break. Perhaps it needs to be a macro that accepts an optional UUID param?

Reply:
Yeah, but we never do that, see: 
- https://github.com/theforeman/smart-proxy/pull/768/files#diff-45f33bea51f0f06fd02907550e854ae2R26
- https://github.com/theforeman/foreman/pull/8034/files#diff-c7642c50c0b5716cfc364d7e3d0b3589R49

Reply:
My problem is that building URLs in the template is a bad pattern and should never be encouraged. We did that in the beginning with `foreman_url`. Then at some point we did need a static parameter. In 170b56c69ab0d7e6b332b352d823bb988eca0b6b we added the parameters option to `foreman_url`. In f7e8785ba53f793ca1ab847bdeb1fa07b77da1d7 we synced the use of it.

I think it may actually make sense to extend `foreman_url` so you would end up using something like `foreman_url('register')`. Then for the user it's a consistent API. We also use the proper Rails routing framework rather than hardcoding routes.

A huge benefit of that is that you can refactor things and user templates do not need to be modified. A user may clone the template to make modifications. At that point upgrading becomes painful. If you use a macro, at least that pain disappears.

In short: it's future proofing your code. Future users and developers will be thankful.

Reply:
We can't use macro here if UUID is only found during the runtime right? Macro is evaluated on the Foreman server before the script gets executed. I agree it's a good practice to avoid this, but how can we achieve it here?

Reply:
You can use the literal `$UUID` as value and rely on shell interpolation (as it does now).

Reply:
> You can use the literal $UUID as value and rely on shell interpolation (as it does now).

Sorry I don't understand what does it mean. Can you explain it please?

Reply:
I guess that means doing something like this

```ruby
register_path :uuid => '$UUID'
```

to render the path, however

1) that would only be applicable to Katello flow, we'd need that as a separate variable, not based on `@registration_url`
2) a bigger problem is, I don't know how to avoid escaping the `$` sign, rails helpers always do it if I'm not mistaken, we'd need to `CGI::unescape(` that part, but that's very ugly. Or use another placeholder that would be replaced with `$UUID` before rendered to the template. Still ugly, but doable, I'll add the code in the controller comment so we can see how it would look like.

I'd prefer to keep appending `?uuid=$UUID`, the generic register url shouldn't really contain parameters. I have another suggestion though for the controller and how we render the URL there.

Reply:
Perhaps the `register_path` macro should just append `?uuid=$UUID` if rails helpers don't work well for our user case? At least abstract it away from the user?